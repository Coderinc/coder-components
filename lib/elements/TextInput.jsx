import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const InputWrapper = styled.div`
  width: 100%;
  padding: 0.5rem 0;

  display: flex;
  flex-wrap: wrap;
`;

const Img = styled.img`
  width: 2rem;
  height: 2rem;
  margin: 0 1rem;
  margin-top: 0.5rem;
`;

const Input = styled.input`
  background-color: transparent;
  caret-color: ${({ theme }) => theme.primary};
  outline: none;
  border: none;
  font-family: circular;
  font-size: ${({ large }) => (large ? '2.25rem' : '0.875rem')};
  font-weight: ${({ bold }) => (bold ? 'bold' : 'lighter')};
  color: ${({ theme }) => theme.primaryFont};
  letter-spacing: 0.0625rem;
  flex: 1;

  &::placeholder {
    color: ${({ theme }) => theme.offsetFont};
  }
`;

const Textarea = styled.textarea`
  background-color: transparent;
  caret-color: ${({ theme }) => theme.primary};
  outline: none;
  border: none;
  font-family: circular;
  font-size: ${({ large }) => (large ? '2.25rem' : '0.875rem')};
  font-weight: ${({ bold }) => (bold ? 'bold' : 'lighter')};
  color: ${({ theme }) => theme.primaryFont};
  letter-spacing: 0.0625rem;
  flex: 1;

  &::placeholder {
    color: ${({ theme }) => theme.offsetFont};
  }
`;

const Error = styled.span`
  color: ${({ theme }) => theme.pink};
  font-size: 0.625rem;
  font-weight: lighter;
  line-height: 1rem;
  width: 100%;
`;

const TextInput = ({
  multiline, icon, meta, ...props
}) => (
  <InputWrapper>
    {icon && <Img src={icon} alt="Icon" />}
    {multiline ? <Textarea {...props} /> : <Input {...props} />}
    <Error>
      {meta.error && meta.touched && meta.visited && !meta.active && meta.error}
    </Error>
  </InputWrapper>
);

TextInput.propTypes = {
  input: PropTypes.object.isRequired,
  meta: PropTypes.object.isRequired,
  large: PropTypes.bool,
  bold: PropTypes.bool,
  multiline: PropTypes.bool,
  icon: PropTypes.string,
};

TextInput.defaultProps = {
  large: false,
  bold: false,
  multiline: false,
  icon: null,
};

export default TextInput;
