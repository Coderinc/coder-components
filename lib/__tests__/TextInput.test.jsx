import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import sinon from 'sinon';
import 'jest-styled-components';
import TextInput from '../elements/TextInput';

describe('TextInput', () => {
  it('input renders correctly', () => {
    const component = renderer.create(<TextInput meta={{}} input={{}} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('textarea renders correctly', () => {
    const component = renderer.create(<TextInput meta={{}} input={{}} multiline />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('fires change on change', () => {
    const onChange = sinon.spy();
    const component = mount(<TextInput meta={{}} input={{ onChange }} />);
    component.find('input').simulate('change');
    expect(onChange.called).toEqual(true);
  });

  it('fires change on change for textarea', () => {
    const onChange = sinon.spy();
    const component = mount(<TextInput meta={{}} input={{ onChange }} multiline />);
    component.find('textarea').simulate('change');
    expect(onChange.called).toEqual(true);
  });
})