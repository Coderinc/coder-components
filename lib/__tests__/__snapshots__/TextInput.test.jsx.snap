// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput input renders correctly 1`] = `
.c0 {
  width: 100%;
  padding: 0.5rem 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c2 {
  font-size: 0.625rem;
  font-weight: lighter;
  line-height: 1rem;
  width: 100%;
}

<div
  className="c0"
>
  <input
    className="c1"
  />
  <span
    className="c2"
  />
</div>
`;

exports[`TextInput textarea renders correctly 1`] = `
.c0 {
  width: 100%;
  padding: 0.5rem 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  background-color: transparent;
  outline: none;
  border: none;
  font-family: circular;
  font-size: 0.875rem;
  font-weight: lighter;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c2 {
  font-size: 0.625rem;
  font-weight: lighter;
  line-height: 1rem;
  width: 100%;
}

<div
  className="c0"
>
  <textarea
    className="c1"
  />
  <span
    className="c2"
  />
</div>
`;
